phac-nml/irida-next
# Metadata: Update values to strings

## What does this PR do and why?
Converting metadata values to strings, so we can search all metadata values. (See #807.)
Fixes #771.

## Screenshots or screen recordings
![image](https://github.com/user-attachments/assets/0496192a-2c05-4cec-8f0d-0134ee7bd423)

## How to set up and validate locally
**Add metadata through the UI:**
1. Navigate to a samples page that the logged in user has access to update.
5. Hit the `Import Metadata` button.
6. Select a file. (See /irida-next-core/test/fixtures/files/metadata/valid.xlsx)
7. Select the `Sample ID` Column.
8. Hit the `Import Metadata` button.
9. Verify the sample metadata was added as strings.
10. Follow the `Test searching metadata` instructions below.

**Add metadata through GraphQL:**
1. Navigate to `http://localhost:3000` and login as admin.
2. Navigate to `http://localhost:3000/graphiql`.
3. Run the following to retrieve a project:
```
query getProject{
  project(fullPath: "bacillus/bacillus-anthracis/outbreak-2022") {
    id
    name
    path
  }
}
```
4. Run the following to create a fresh sample:
```
mutation createSample{
  createSample(input: { 
    projectId: "<insert here>", 
    name: "Sample Name", 
    description: "Sample Description" 
  }) {
    sample {
      id,
      puid,
      name,
      description
    }
    errors {
      path,
      message
    }
  }
}
```
5. Run the following to add metadata to the sample:
```
mutation addMetadataToSample{
  updateSampleMetadata(input: {
    sampleId: "<insert here>", 
    metadata: { true_boolean: true, false_boolean: false, nil: null, empty: "", number: 123, date: "2024-03-11" }
  }) {
    sample{
      id,
      name,
      description,
      metadata
    },
    status,
    errors {
      path
      message
    }
  }
}
```
6. Verify the date, number, true_boolean, and false_boolean were added successfully as strings.
7. Follow the `Test searching metadata` instructions below.

**Test searching metadata**
1. Navigate to a samples page.
2. Toggle metadata.
3. Search using the format FIELD_NAME:FIELD_VALUE.
4. Verify the results when searching for dates, numbers, booleans, etc. 
 
## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
